rfd/0083-machine-id-host-certs.md

```diff
@@ -0,0 +1,198 @@
+---
+authors: Tim Buckley (tim@goteleport.com)
+state: draft
+---
+
+# RFD 83 - Host Certificate support for Machine ID
+
+## What
+
+Allow Machine ID bots to issue host certificates for OpenSSH servers, per our
+documented [manual instructions].
+
+This has two components:
+1. Allow bots to securely issue host certificates for particular hosts via RBAC
+   rules
+2. Provide some bot UX for issuing and making use of issued host certificates 
+
+[manual instructions]: https://goteleport.com/docs/server-access/guides/openssh/#step-24-configure-host-authentication
+
+## Why
+
+Issuing host certs today requires someone with admin-level permissions to
+manually issue certificates via `tctl auth sign ...`. If these certificates
+expire or if the cluster CAs are rotated, these certificates need to be manually
+refreshed. This is inconvenient and encourages users to distribute long lived
+credentials when onboarding OpenSSH nodes.
+
+## Details
+
+### Issuing host certificates securely
+
+In Teleport today, host certificates are not modelled as a real resource.
+However, we do expose a `host_cert` RBAC resource. Any user granted the `create`
+verb on this (virtual) resource may call `GenerateHostCert()` with any
+parameters they wish and receive a set of host certificates.
+
+This is reasonable for human cluster administrators with short-lived credentials
+protected by MFA, but problematic for long-lived machine credentials. Stolen
+credentials, even if short lived, could be used to issue additional long-lived
+certificates. While these host credentials don't provide a large level of access
+to the Teleport cluster, they could be used to impersonate real nodes.
+
+To mitigate this, we'd like to implement support for `where` clauses for
+`host_cert` RBAC rules as we do for rules backed by real Teleport resources. For
+example, this role would allow a user to issue host certs, but only for
+principals ending in `foo.example.com`:
+
+```yaml
+kind: role
+version: v5
+metadata:
+  name: example
+spec:
+  allow:
+    rules:
+      - resources: [host_cert]
+        verbs: [create]
+        where: ends_with(host_cert.principals, ".foo.example.com")
+```
+
+This involves:
+1. Adding new comparison functions to better match against DNS names. Minimally,
+   we'll add an `ends_with(inputs, suffix)`, but may consider a regex or glob
+   matcher. The latter may have problematic quoting and escaping requirements
+   that may not be worth solving in the underlying `predicate` library, so we'll
+   avoid implementing regex matchers until a meaningful use-case exists.
+
+2. Ensuring these functions support lists of strings (or building separate
+   functions specifically for matching against `[]string`).
+
+   For example, a hypothetical `ends_with(principals, ".example.com")` should 
```

Comment:
Can you think of any valid use cases for an OR instead of an AND?

I ask because we're getting lots of requests to support more OR conditions in our RBAC engine lately.

We could instead have `some_end_width(principals, suffix)` and `all_end_with(principals, suffix)` if we need greater flexibility.

Reply:
That's a good thought, and that naming scheme opens the door for a few extra functions like `all_equal()`. That said, I think `some_end_with()` would probably be pretty dangerous in most RBAC situations since it effectively allows anything through.

We don't have a great syntax for "allow any subset of these options" right now. It is possible (albeit ugly) to use something like:

```
all_end_with(host_cert.principals, ".foo.example.com") || all_end_with(host_cert.principals, ".bar.example.com")
```

... but this wouldn't allow requesting _both_ `node.foo.example.com` and `node.bar.example.com`. Ultimately we'd need to specify a list / varargs and I don't believe the `predicate` library has any support for that syntactically. I'm not sure there's a great way forward here aside from a regex matcher (which has its own problems).